### PR TITLE
[SPARK-43128][CONNECT][SS] Make `recentProgress` and `lastProgress` return `StreamingQueryProgress` consistent with the native Scala Api

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr  add support to make `recentProgress` and `lastProgress` in `RemoteStreamingQuery` return `StreamingQueryProgress` instance consistent with the native Scala Api.

### Why are the changes needed?
Add Spark connect jvm client api coverage.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new check to `StreamingQuerySuite`

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: `observedMetrics` maybe null, so add `@JsonSetter(nulls = Nulls.AS_EMPTY)` to set it as empty map when it is null

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:56]: Should we also add `@JsonSetter(nulls = Nulls.AS_EMPTY)` as protection for other collection types?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:44]: Just used in `[spark]` , So no mima check was added

**Comment by LuciferYang** [comment]
If this is not consistent with the initial idea of this ticket, please let me know

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala
Comment (bogao007) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:162]: Aggregated metrics like `numInputRows`, `inputRowsPerSecond` and `processedRowsPerSecond` are being calculated inside this class. Can they be safely deserialized from json string? Might also worth to add them into the test case.

**Comment by bogao007** [comment]
cc @rangadi

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala
Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Deserialize result of `observedMetrics` has problem, `scheam` fo GenericRowWithSchema is null

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: https://github.com/apache/spark/blob/d407a42090d7c027050be7ee723f7e3d8f686ed7/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala#L2182-L2186

@rangadi @bogao007 I think we should serialize the `StreamingQueryProgress` to json as a pojo to  maintain the row schema information in `observedMetrics`, `progressReports.map(_.json)` will lose schmea information of `GenericRowWithSchema`, this results in the inability to deserialize to GenericRowWithSchema on the client side

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Maybe py test will fail, let's wait for ci

Comment (bogao007) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/progress.scala:158]: Yep, we would not be able to deserialize `observedMetrics` without this change, thanks for the update!

**Review by bogao007 — APPROVED** [review]
LGTM overall, with one minor comment on the test case. Please also get an approval from @rangadi before merging, thanks!

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryProgressSuite.scala
Comment (LuciferYang) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryProgressSuite.scala]: Added a `sources` item, there change to `assert(result.numInputRows === 1467) // 678 + 789`

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQuery.scala
Comment (rangadi) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/streaming/StreamingQuery.scala:173]: This is a good idea (at least for now). So essentially we don't need to add many protobuf message to match :).

**Review by rangadi — APPROVED** [review]
Thanks for adding this!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by LuciferYang** [comment]
rebased and resolved conflicts

**Comment by LuciferYang** [comment]
friendly ping @HyukjinKwon can we merge this one?

**Comment by HeartSaVioR** [comment]
I'll leave this to you for self-merging, so that you can test your new permission. Congrats again :)

**Comment by LuciferYang** [comment]
> I'll leave this to you for self-merging, so that you can test your new permission. Congrats again :)

Thanks @HeartSaVioR :)

**Comment by HeartSaVioR** [comment]
I'll just help merging this one as it has been here for multiple weeks and we don't want to require this PR to be rebased anymore.

**Comment by HeartSaVioR** [comment]
Thanks @LuciferYang , I merged this to master.

**Pull request closed without merging**
